File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/model/network/method/message/data/lao/StateLaoTest.java
package com.github.dedis.popstellar.model.network.method.message.data.lao;

import androidx.test.ext.junit.runners.AndroidJUnit4;

import com.github.dedis.popstellar.model.network.JsonTestUtils;
import com.github.dedis.popstellar.model.network.method.message.PublicKeySignaturePair;
import com.github.dedis.popstellar.model.network.method.message.data.Action;
import com.github.dedis.popstellar.model.network.method.message.data.Objects;
import com.github.dedis.popstellar.model.objects.Lao;
import com.github.dedis.popstellar.model.objects.security.MessageID;
import com.github.dedis.popstellar.model.objects.security.PublicKey;
import com.github.dedis.popstellar.testutils.Base64DataUtils;
import com.google.gson.JsonParseException;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.internal.util.collections.Sets;

import java.time.Instant;
import java.util.*;

import static com.github.dedis.popstellar.testutils.Base64DataUtils.*;
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.junit.Assert.*;

@RunWith(AndroidJUnit4.class)
public class StateLaoTest {

  private final String name = " Lao name";
  private final long creation = Instant.now().getEpochSecond() - 10000;
  private final long lastModified = creation + 1;
  private final PublicKey organizer = generatePublicKey();
  private final MessageID modificationId = generateMessageID();
  private final Set<PublicKey> witnesses = Sets.newSet(generatePublicKey(), generatePublicKey());
  private final String id = Lao.generateLaoId(organizer, creation, name);
  private final List<PublicKeySignaturePair> modificationSignatures =
      Collections.singletonList(
          new PublicKeySignaturePair(generatePublicKey(), generateSignature()));
  private final StateLao stateLao =
      new StateLao(
          id,
          name,
          creation,
          lastModified,
          organizer,
          modificationId,
          witnesses,
          modificationSignatures);

  @Test(expected = IllegalArgumentException.class)
  public void constructorFailsIdNotBase64Test() {
    new StateLao(
        "wrong id",
        name,
        creation,
        lastModified,
        organizer,
        modificationId,
        witnesses,
        modificationSignatures);
  }

  @Test(expected = IllegalArgumentException.class)
  public void constructorFailsInvalidIdHashTest() {
    String wrongId = "A" + id.substring(1);
    new StateLao(
        wrongId,
        name,
        creation,
        lastModified,
        organizer,
        modificationId,
        witnesses,
        modificationSignatures);
  }

  @Test(expected = IllegalArgumentException.class)
  public void constructorFailsEmptyNameTest() {
    new StateLao(
        id,
        "",
        creation,
        lastModified,
        organizer,
        modificationId,
        witnesses,
        modificationSignatures);
  }

  @Test(expected = IllegalArgumentException.class)
  public void constructorFailsFutureCreationTimeTest() {
    long futureCreation = Instant.now().getEpochSecond() + 1000;
    new StateLao(
        id,
        name,
        futureCreation,
        lastModified,
        organizer,
        modificationId,
        witnesses,
        modificationSignatures);
  }

  @Test(expected = IllegalArgumentException.class)
  public void constructorFailsFutureModificationTimeTest() {
    long futureModification = Instant.now().getEpochSecond() + 1000;
    new StateLao(
        id,
        name,
        creation,
        futureModification,
        organizer,
        modificationId,
        witnesses,
        modificationSignatures);
  }

  @Test(expected = IllegalArgumentException.class)
  public void constructorFailsModificationBeforeCreationTimeTest() {
    new StateLao(
        id,
        name,
        creation,
        creation - 10,
        organizer,
        modificationId,
        witnesses,
        modificationSignatures);
  }

  @Test
  public void getIdTest() {
    assertThat(stateLao.getId(), is(id));
  }

  @Test
  public void getObjectTest() {
    assertThat(stateLao.getObject(), is(Objects.LAO.getObject()));
  }

  @Test
  public void getActionTest() {
    assertThat(stateLao.getAction(), is(Action.STATE.getAction()));
  }

  @Test
  public void getNameTest() {
    assertThat(stateLao.getName(), is(name));
  }

  @Test
  public void getCreationTest() {
    assertThat(stateLao.getCreation(), is(creation));
  }

  @Test
  public void getLastModifiedTest() {
    assertThat(stateLao.getLastModified(), is(lastModified));
  }

  @Test
  public void getOrganizerTest() {
    assertThat(stateLao.getOrganizer(), is(organizer));
  }

  @Test
  public void getWitnessesTest() {
    assertThat(stateLao.getWitnesses(), is(witnesses));
  }

  @Test
  public void getModificationIdTest() {
    assertThat(stateLao.getModificationId(), is(modificationId));
  }

  @Test
  public void getModificationIdSignaturesTest() {
    assertThat(stateLao.getModificationId(), is(modificationId));
  }

  @Test
  public void isEqualTest() {
    assertEquals(
        stateLao,
        new StateLao(
            id,
            name,
            creation,
            lastModified,
            organizer,
            modificationId,
            witnesses,
            modificationSignatures));
    // The modification id isn't taken into account to know if they are equal
    assertEquals(
        stateLao,
        new StateLao(
            id,
            name,
            creation,
            lastModified,
            organizer,
            Base64DataUtils.generateMessageIDOtherThan(modificationId),
            witnesses,
            modificationSignatures));
    // same goes for modification signatures
    assertEquals(
        stateLao,
        new StateLao(id, name, creation, lastModified, organizer, modificationId, witnesses, null));
    String random = " random string";
    String newId = Lao.generateLaoId(organizer, creation, random);
    assertNotEquals(
        stateLao,
        new StateLao(
            newId,
            random,
            creation,
            lastModified,
            organizer,
            modificationId,
            witnesses,
            modificationSignatures));
    PublicKey newKey = Base64DataUtils.generatePublicKeyOtherThan(organizer);
    newId = Lao.generateLaoId(newKey, creation, name);
    assertNotEquals(
        stateLao,
        new StateLao(
            newId,
            name,
            creation,
            lastModified,
            newKey,
            modificationId,
            witnesses,
            modificationSignatures));
    newId = Lao.generateLaoId(organizer, creation - 1, name);
    assertNotEquals(
        stateLao,
        new StateLao(
            newId,
            name,
            creation - 1,
            lastModified,
            organizer,
            modificationId,
            witnesses,
            modificationSignatures));
    assertNotEquals(
        stateLao,
        new StateLao(
            id,
            name,
            creation,
            creation + 10,
            organizer,
            modificationId,
            witnesses,
            modificationSignatures));
    assertNotEquals(
        stateLao,
        new StateLao(
            id,
            name,
            creation,
            lastModified,
            organizer,
            modificationId,
            Sets.newSet(generatePublicKey()),
            modificationSignatures));
  }

  @Test
  public void jsonValidationTest() {
    JsonTestUtils.testData(stateLao);

    String pathDir = "protocol/examples/messageData/lao_state/";
    String jsonInvalid1 =
        JsonTestUtils.loadFile(pathDir + "wrong_lao_state_additional_params.json");
    String jsonInvalid2 = JsonTestUtils.loadFile(pathDir + "wrong_lao_state_missing_params.json");

    String jsonInvalid3 = JsonTestUtils.loadFile(pathDir + "bad_lao_state_creation_negative.json");
    String jsonInvalid4 =
        JsonTestUtils.loadFile(pathDir + "bad_lao_state_last_modified_negative.json");
    String jsonInvalid5 =
        JsonTestUtils.loadFile(pathDir + "bad_lao_state_organizer_not_base64.json");
    String jsonInvalid6 = JsonTestUtils.loadFile(pathDir + "bad_lao_state_witness_not_base64.json");

    assertThrows(JsonParseException.class, () -> JsonTestUtils.parse(jsonInvalid1));
    assertThrows(JsonParseException.class, () -> JsonTestUtils.parse(jsonInvalid2));
    assertThrows(JsonParseException.class, () -> JsonTestUtils.parse(jsonInvalid3));
    assertThrows(JsonParseException.class, () -> JsonTestUtils.parse(jsonInvalid4));
    assertThrows(JsonParseException.class, () -> JsonTestUtils.parse(jsonInvalid5));
    assertThrows(JsonParseException.class, () -> JsonTestUtils.parse(jsonInvalid6));
  }
}

File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/model/network/method/message/data/lao/StateLaoTest.kt
package com.github.dedis.popstellar.model.network.method.message.data.lao

import androidx.test.ext.junit.runners.AndroidJUnit4
import com.github.dedis.popstellar.model.network.JsonTestUtils.loadFile
import com.github.dedis.popstellar.model.network.JsonTestUtils.parse
import com.github.dedis.popstellar.model.network.JsonTestUtils.testData
import com.github.dedis.popstellar.model.network.method.message.PublicKeySignaturePair
import com.github.dedis.popstellar.model.network.method.message.data.Action
import com.github.dedis.popstellar.model.network.method.message.data.Objects
import com.github.dedis.popstellar.model.objects.Lao.Companion.generateLaoId
import com.github.dedis.popstellar.testutils.Base64DataUtils
import com.google.gson.JsonParseException
import java.time.Instant
import org.hamcrest.CoreMatchers
import org.hamcrest.MatcherAssert
import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.internal.util.collections.Sets

@RunWith(AndroidJUnit4::class)
class StateLaoTest {
  private val name = " Lao name"
  private val creation = Instant.now().epochSecond - 10000
  private val lastModified = creation + 1
  private val organizer = Base64DataUtils.generatePublicKey()
  private val modificationId = Base64DataUtils.generateMessageID()
  private val witnesses =
    Sets.newSet(Base64DataUtils.generatePublicKey(), Base64DataUtils.generatePublicKey())
  private val id = generateLaoId(organizer, creation, name)
  private val modificationSignatures =
    listOf(
      PublicKeySignaturePair(
        Base64DataUtils.generatePublicKey(),
        Base64DataUtils.generateSignature()
      )
    )
  private val stateLao =
    StateLao(
      id,
      name,
      creation,
      lastModified,
      organizer,
      modificationId,
      witnesses,
      modificationSignatures
    )

  @Test(expected = IllegalArgumentException::class)
  fun constructorFailsIdNotBase64Test() {
    StateLao(
      "wrong id",
      name,
      creation,
      lastModified,
      organizer,
      modificationId,
      witnesses,
      modificationSignatures
    )
  }

  @Test(expected = IllegalArgumentException::class)
  fun constructorFailsInvalidIdHashTest() {
    val wrongId = "A" + id.substring(1)
    StateLao(
      wrongId,
      name,
      creation,
      lastModified,
      organizer,
      modificationId,
      witnesses,
      modificationSignatures
    )
  }

  @Test(expected = IllegalArgumentException::class)
  fun constructorFailsEmptyNameTest() {
    StateLao(
      id,
      "",
      creation,
      lastModified,
      organizer,
      modificationId,
      witnesses,
      modificationSignatures
    )
  }

  @Test(expected = IllegalArgumentException::class)
  fun constructorFailsFutureCreationTimeTest() {
    val futureCreation = Instant.now().epochSecond + 1000
    StateLao(
      id,
      name,
      futureCreation,
      lastModified,
      organizer,
      modificationId,
      witnesses,
      modificationSignatures
    )
  }

  @Test(expected = IllegalArgumentException::class)
  fun constructorFailsFutureModificationTimeTest() {
    val futureModification = Instant.now().epochSecond + 1000
    StateLao(
      id,
      name,
      creation,
      futureModification,
      organizer,
      modificationId,
      witnesses,
      modificationSignatures
    )
  }

  @Test(expected = IllegalArgumentException::class)
  fun constructorFailsModificationBeforeCreationTimeTest() {
    StateLao(
      id,
      name,
      creation,
      creation - 10,
      organizer,
      modificationId,
      witnesses,
      modificationSignatures
    )
  }

  @Test
  fun idTest() {
    MatcherAssert.assertThat(stateLao.id, CoreMatchers.`is`(id))
  }

  @Test
  fun objectTest() {
    MatcherAssert.assertThat(stateLao.`object`, CoreMatchers.`is`(Objects.LAO.`object`))
  }

  @Test
  fun actionTest() {
    MatcherAssert.assertThat(stateLao.action, CoreMatchers.`is`(Action.STATE.action))
  }

  @Test
  fun nameTest() {
    MatcherAssert.assertThat(stateLao.name, CoreMatchers.`is`(name))
  }

  @Test
  fun creationTest() {
    MatcherAssert.assertThat(stateLao.creation, CoreMatchers.`is`(creation))
  }

  @Test
  fun lastModifiedTest() {
    MatcherAssert.assertThat(stateLao.lastModified, CoreMatchers.`is`(lastModified))
  }

  @Test
  fun organizerTest() {
    MatcherAssert.assertThat(stateLao.organizer, CoreMatchers.`is`(organizer))
  }

  @Test
  fun witnessesTest() {
    MatcherAssert.assertThat(stateLao.witnesses, CoreMatchers.`is`(witnesses))
  }

  @Test
  fun modificationIdTest() {
    MatcherAssert.assertThat(stateLao.modificationId, CoreMatchers.`is`(modificationId))
  }

  @Test
  fun modificationIdSignaturesTest() {
    MatcherAssert.assertThat(stateLao.modificationId, CoreMatchers.`is`(modificationId))
  }

  @Test
  fun isEqualTest() {
    Assert.assertEquals(
      stateLao,
      StateLao(
        id,
        name,
        creation,
        lastModified,
        organizer,
        modificationId,
        witnesses,
        modificationSignatures
      )
    )
    // The modification id isn't taken into account to know if they are equal
    Assert.assertEquals(
      stateLao,
      StateLao(
        id,
        name,
        creation,
        lastModified,
        organizer,
        Base64DataUtils.generateMessageIDOtherThan(modificationId),
        witnesses,
        modificationSignatures
      )
    )
    // same goes for modification signatures
    Assert.assertEquals(
      stateLao,
      StateLao(id, name, creation, lastModified, organizer, modificationId, witnesses, null)
    )
    val random = " random string"
    var newId = generateLaoId(organizer, creation, random)
    Assert.assertNotEquals(
      stateLao,
      StateLao(
        newId,
        random,
        creation,
        lastModified,
        organizer,
        modificationId,
        witnesses,
        modificationSignatures
      )
    )
    val newKey = Base64DataUtils.generatePublicKeyOtherThan(organizer)
    newId = generateLaoId(newKey, creation, name)
    Assert.assertNotEquals(
      stateLao,
      StateLao(
        newId,
        name,
        creation,
        lastModified,
        newKey,
        modificationId,
        witnesses,
        modificationSignatures
      )
    )
    newId = generateLaoId(organizer, creation - 1, name)
    Assert.assertNotEquals(
      stateLao,
      StateLao(
        newId,
        name,
        creation - 1,
        lastModified,
        organizer,
        modificationId,
        witnesses,
        modificationSignatures
      )
    )
    Assert.assertNotEquals(
      stateLao,
      StateLao(
        id,
        name,
        creation,
        creation + 10,
        organizer,
        modificationId,
        witnesses,
        modificationSignatures
      )
    )
    Assert.assertNotEquals(
      stateLao,
      StateLao(
        id,
        name,
        creation,
        lastModified,
        organizer,
        modificationId,
        Sets.newSet(Base64DataUtils.generatePublicKey()),
        modificationSignatures
      )
    )
  }

  @Test
  fun jsonValidationTest() {
    testData(stateLao)
    val pathDir = "protocol/examples/messageData/lao_state/"
    val jsonInvalid1 = loadFile(pathDir + "wrong_lao_state_additional_params.json")
    val jsonInvalid2 = loadFile(pathDir + "wrong_lao_state_missing_params.json")
    val jsonInvalid3 = loadFile(pathDir + "bad_lao_state_creation_negative.json")
    val jsonInvalid4 = loadFile(pathDir + "bad_lao_state_last_modified_negative.json")
    val jsonInvalid5 = loadFile(pathDir + "bad_lao_state_organizer_not_base64.json")
    val jsonInvalid6 = loadFile(pathDir + "bad_lao_state_witness_not_base64.json")

    Assert.assertThrows(JsonParseException::class.java) { parse(jsonInvalid1) }
    Assert.assertThrows(JsonParseException::class.java) { parse(jsonInvalid2) }
    Assert.assertThrows(JsonParseException::class.java) { parse(jsonInvalid3) }
    Assert.assertThrows(JsonParseException::class.java) { parse(jsonInvalid4) }
    Assert.assertThrows(JsonParseException::class.java) { parse(jsonInvalid5) }
    Assert.assertThrows(JsonParseException::class.java) { parse(jsonInvalid6) }
  }
}
